src/OmniSharp.Roslyn.CSharp/Services/Decompilation/OmniSharpCSharpDecompiledSourceService.cs

```diff
@@ -0,0 +1,191 @@
+    {
+        private readonly HostLanguageServices _provider;
+        private readonly ILoggerFactory _loggerFactory;
+        private const string MetadataAsSourceHelpers = "Microsoft.CodeAnalysis.MetadataAsSource.MetadataAsSourceHelpers";
+        private const string CSharpDocumentationCommentFormattingService = "Microsoft.CodeAnalysis.CSharp.DocumentationComments.CSharpDocumentationCommentFormattingService";
+        private const string DocCommentConverter = "Microsoft.CodeAnalysis.CSharp.DocumentationComments.DocCommentConverter";
+        private static readonly FileVersionInfo decompilerVersion = FileVersionInfo.GetVersionInfo(typeof(CSharpDecompiler).Assembly.Location);
+        private readonly Lazy<Assembly> _roslynFeatureAssembly;
+        private readonly Lazy<Assembly> _csharpFeatureAssembly;
+        private readonly Lazy<Type> _csharpMetadataAsSourceService;
+        private readonly Lazy<Type> _csharpDocumentationCommentFormattingService;
+        private readonly Lazy<Type> _docCommentConverter;
+        private readonly Lazy<MethodInfo> _metadataGetAssemblyInfo;
+        private readonly Lazy<MethodInfo> _metadataGetAssemblyDisplay;
+
+        public OmniSharpCSharpDecompiledSourceService(HostLanguageServices provider, IAssemblyLoader loader, ILoggerFactory loggerFactory)
+        {
+            _roslynFeatureAssembly = loader.LazyLoad(Configuration.RoslynFeatures);
+            _csharpFeatureAssembly = loader.LazyLoad(Configuration.RoslynCSharpFeatures);
+            _csharpMetadataAsSourceService = _roslynFeatureAssembly.LazyGetType(MetadataAsSourceHelpers);
+            _csharpDocumentationCommentFormattingService = _csharpFeatureAssembly.LazyGetType(CSharpDocumentationCommentFormattingService);
+            _docCommentConverter = _csharpFeatureAssembly.LazyGetType(DocCommentConverter);
+            _metadataGetAssemblyInfo = _csharpMetadataAsSourceService.LazyGetMethod("GetAssemblyInfo");
+            _metadataGetAssemblyDisplay = _csharpMetadataAsSourceService.LazyGetMethod("GetAssemblyDisplay");
+
+            _provider = provider;
+            _loggerFactory = loggerFactory;
+        }
+
+        public async Task<Document> AddSourceToAsync(Document document, Compilation symbolCompilation, Microsoft.CodeAnalysis.ISymbol symbol, CancellationToken cancellationToken)
+        {
+            // Get the name of the type the symbol is in
+            var containingOrThis = symbol.GetContainingTypeOrThis();
+            var fullName = GetFullReflectionName(containingOrThis);
+
+            string assemblyLocation = null;
+            var isReferenceAssembly = symbol.ContainingAssembly.GetAttributes().Any(attribute => attribute.AttributeClass.Name == nameof(ReferenceAssemblyAttribute)
+                && attribute.AttributeClass.ToNameDisplayString() == typeof(ReferenceAssemblyAttribute).FullName);
+            if (isReferenceAssembly)
+            {
+                try
```

Comment:
Does this logic let you resolve the contents of ref assemblies to their runtime equivalents?  If so I'll port this over to ionide, because we currently have this same problem with resolving ref assemblies instead of implementation assemblies.

Reply:
it will only work when the implementation is in the GAC (this is the same as in VS)

Reply:
Gotcha, so for .Net Core tooing you'd need to find the matching non-ref assembly locations and decompile those instead. About what I figured :) Thanks for clarifying.